File: modules/client/room.cc
/* 
 * Copyright (C) 2017 Charybdis Development Team
 * Copyright (C) 2017 Jason Volk <jason@zemos.net>
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice is present in all copies.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
 * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

using namespace ircd;

const database::descriptor room_created_descriptor
{
	"created",

	R"(### developer note:
	A UNIX epoch timestamp sampled when the room was created.
	)",
	{
		// readable key      // binary value
		typeid(string_view), typeid(time_t)
	}
};

const database::descriptor room_creator_descriptor
{
	"creator",

	R"(### protocol note:
	The user_id of the room creator. This is set by the homeserver.
	)"
};

const database::descriptor room_topic_descriptor
{
	"topic",

	R"(### protocol note:
	If this is included, an m.room.topic event will be sent into the room to indicate the
	topic for the room. See Room Events for more information on m.room.topic.
	)"
};

const database::descriptor room_visibility_descriptor
{
	"visibility",

	R"(### protocol note:
	Rooms default to private visibility if this key is not included.

	* "public" visibility indicates that the room will be shown in the published room list.

	* "private" visibility will hide the room from the published room list.

	One of: ["public", "private"]
	)"
};

const database::descriptor room_join_rules_descriptor
{
	"join_rules",

	R"(### protocol note:
	A room may be public meaning anyone can join the room without any prior action.
	Alternatively, it can be invite meaning that a user who wishes to join the room must first
	receive an invite to the room from someone already inside of the room. Currently, knock and private
	are reserved keywords which are not implemented.

	The type of rules used for users wishing to join this room.
	One of: ["public", "knock", "invite", "private"]
	)"
};

const database::descriptor room_history_visibility_descriptor
{
	"history_visibility",

	R"(### protocol note:

	* "world_readable" All events while this is the m.room.history_visibility value may be shared
	by any participating homeserver with anyone, regardless of whether they have ever joined the room.

	* "shared" Previous events are always accessible to newly joined members. All events in the room
	are accessible, even those sent when the member was not a part of the room.

	* "invited" Events are accessible to newly joined members from the point they were invited onwards.
	Events stop being accessible when the member's state changes to something other than invite or join.

	* "joined" Events are accessible to newly joined members from the point they joined the room onwards.
	Events stop being accessible when the member's state changes to something other than join.
	)"
};

const database::descriptor room_alias_descriptor
{
	"alias",

	R"(### protocol note:
	The desired room alias local part. If this is included, a room alias will be created and mapped to
	the newly created room. The alias will belong on the same homeserver which created the room.
	For example, if this was set to "foo" and sent to the homeserver "example.com" the complete
	room alias would be #foo:example.com.

	### developer note:
	The alias column on the room's primary row has a comma separated list of aliases.
	For each of those aliases this column has a key indexed for it; the value for that key is
	the primary room's name. This is a cross-reference that must be kept in sync.
	)"
};

const database::descriptor room_federate_descriptor
{
	"federate",

	R"(### protocol note:
	Whether users on other servers can join this room. Defaults to true if key does not exist.
	)",
	{
		// readable key      // binary value
		typeid(string_view), typeid(bool)
	}
};

const database::description room_description
{
	{ "default" },
	room_created_descriptor,
	room_creator_descriptor,
	room_topic_descriptor,
	room_visibility_descriptor,
	room_join_rules_descriptor,
	room_history_visibility_descriptor,
	room_alias_descriptor,
	room_federate_descriptor,
};

std::shared_ptr<database> room_database
{
	std::make_shared<database>("room"s, ""s, room_description)
};

extern database *const room
{
	room_database.get()
};

struct room
:resource
{
	using resource::resource;
}
room_resource
{
	"_matrix/client/r0/room",
	"Rooms (7.0)"
};

mapi::header IRCD_MODULE
{
	"registers the resource 'client/room' to manage Matrix rooms"
};

File: modules/static/charybdis/room.js
/*
 * IRCd Charybdis 5/Matrix
 *
 * Copyright (C) 2017 Charybdis Development Team
 * Copyright (C) 2017 Jason Volk (jason@zemos.net)
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice is present in all copies.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
 * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 */

'use strict';

/**
 ******************************************************************************
 * Room
 *
 */
mc.room = class
{
	constructor(summary = {}, opts = {})
	{
		// Access/convenience to user's options structure
		this.opts = opts;
		Object.defaults(this.opts, mc.room.opts);
		if(maybe(() => summary.room_id.endsWith(":localhost")))
			this.opts.local = true;

		if(!Array.isArray(summary))
			this.id = summary.room_id;
		else
			this.id = summary[0].room_id;

		// Identity
		this.sid = mc.m.sid(this.id);
		this.uri = encodeURIComponent(this.id);

		// TODO: purge
		this.control = Object.copy(room.control);

		// Persistent storage
		if(!(this.id in mc.local.room)) mc.local.room[this.id] = {};
		if(!(this.id in mc.session.room)) mc.session.room[this.id] = {};
		if(!(this.id in mc.instance.room)) mc.instance.room[this.id] = {};

		this.local = mc.local.room[this.id];
		this.session = mc.session.room[this.id];
		this.instance = mc.instance.room[this.id];

		// Event Machine
		this.timeline = new room.timeline;
		this.timeline.id = this.id; //tmp

		this.oracle = new room.state(this.timeline);
		this.state = new Proxy(this.oracle.cache, this.oracle);
		this.content = new Proxy(this.state, this.oracle.content);

		// Accessories
		this.notifications = Function.bindtree(room.notifications, this);
		this.account = Function.bindtree(room.account, this);
		this.power = Function.bindtree(room.power, this);
		this.membership = Function.bindtree(room.membership, this, summary);
		this.input = Function.bindtree(room.input, this);
		this.dom = Function.bindtree(room.dom, this);
		this.receipt = Function.bindtree(room.receipt, this);
		this.typing = Function.bindtree(room.typing, this);
		this.focus = Function.bindtree(room.focus, this);
		this.scroll = Function.bindtree(room.scroll, this);
		this.sync = Function.bindtree(room.sync, this);

		// Generate some initial pseudo-state from the summary data.
		if(!Array.isArray(summary))
			this.timeline.insert(room.state.summary.parse(summary));
		else
			this.timeline.insert(summary);
	}

	get summary()
	{
		return room.state.summary(this.state);
	}
};

// Global scope alias
var room = mc.room;

/**
 * Default Room options
 *
 * Items in here only change when the user specifically changes settings;
 * unlike the control structure below which holds any misc model state.
 */
room.opts =
{
};

/**
 * Default Misc model state
 *
 * These things should be categorized better
 */
room.control =
{
	show_members: false,
	show_main_input: true,
	show_status_frieze: true,
	show_event_detail: false,
	show_event_info: {},

	left:
	{
		member: {},
	},

	input: "",            // ng-model of the main input textarea
	files: {},            // name => FileReader attached alongside the text
	evalmode: false,      // true = input will not be sent to channel

	content:
	{
		type: {},         // 'm.room.member': false,
		event_id: {},     // '$xyz:foo.com': false,
		zoom: undefined,  // event_id
	},

	messages: {},
};
